program/rust/src/processor/upd_price.rs

```diff
@@ -175,21 +176,82 @@ pub fn upd_price(
+                cmd_args.price,
+                cmd_args.confidence,
+                cmd_args.status,
+            )?;
+
+            {
+                let publisher_price = &mut price_data.comp_[publisher_index].latest_;
+                publisher_price.price_ = cmd_args.price;
+                publisher_price.conf_ = cmd_args.confidence;
+                publisher_price.status_ = status;
+                publisher_price.pub_slot_ = cmd_args.publishing_slot;
+            }
+        }
+    }
 
-    // NOTE: c_upd_aggregate must use a raw pointer to price
-    // data. Solana's `<account>.borrow_*` methods require exclusive
-    // access, i.e. no other borrow can exist for the account.
-    #[allow(unused_assignments)]
-    if clock.slot > latest_aggregate_price.pub_slot_ {
+    let should_call_c_upd_aggregate: bool = if cfg!(feature = "pythnet") {
+        clock.slot >= latest_aggregate_price.pub_slot_
+    } else {
+        clock.slot > latest_aggregate_price.pub_slot_
+    };
+
+    let aggregate_updated = if should_call_c_upd_aggregate {
         unsafe {
-            aggregate_updated = c_upd_aggregate(
+            c_upd_aggregate(
                 price_account.try_borrow_mut_data()?.as_mut_ptr(),
                 clock.slot,
                 clock.unix_timestamp,
+            )
+        }
+    } else {
+        false // Do not call c_upd_aggregate and set aggregate_updated to false
+    };
+
+    // get number of slots from last update
+    let slots_since_last_update = clock.slot - latest_aggregate_price.pub_slot_;
+
+    if aggregate_updated {
+        // The price_data borrow happens in a scope because it must be
+        // dropped before we borrow again as raw data pointer for the C
+        // aggregation logic.
+        #[cfg(feature = "pythnet")]
+        {
+            // get price data here
+            let mut price_data =
+                load_checked::<PriceAccount>(price_account, cmd_args.header.version)?;
+            // check if its the first price update in the slot
+            if slots_since_last_update > 0 {
+                price_data.prev_twap_ = price_data.twap_;
+                price_data.prev_twac_ = price_data.twac_;
+                price_data.prev_price_cumulative = price_data.price_cumulative;
+            }
+            price_data.twap_ = price_data.prev_twap_;
+            price_data.twac_ = price_data.prev_twac_;
+            price_data.price_cumulative = price_data.prev_price_cumulative;
+        }
+        unsafe {
+            c_upd_twap(
+                price_account.try_borrow_mut_data()?.as_mut_ptr(),
+                slots_since_last_update as i64, // Ensure slots_since_last_update is cast to i64, as expected by the function signature
             );
+            #[cfg(feature = "pythnet")]
+            {
+                let mut price_data =
+                    load_checked::<PriceAccount>(price_account, cmd_args.header.version)?;
```

Comment:
Maybe you can move this to the block that starts in line 229 so it's more readable